src/dmd/clone.d

```diff
@@ -765,7 +765,10 @@ extern (C++) FuncDeclaration buildXtoHash(StructDeclaration sd, Scope* sc)
     const(char)* code =
         "size_t h = 0;" ~
         "foreach (i, T; typeof(p.tupleof))" ~
-        "    h = h * 33 + typeid(T).getHash(cast(const void*)&p.tupleof[i]);" ~
+        "    static if(is(T* : const(Object)*)) " ~
+        "        h = h * 33 + typeid(const(Object)).getHash(cast(const void*)&p.tupleof[i]);" ~
+        "    else " ~
+        "        h = h * 33 + typeid(T).getHash(cast(const void*)&p.tupleof[i]);" ~
```

Comment:
OT: We should look at whether Knuth hashing is actually a sane choice here.
[Multiplicative hashing](https://en.wikipedia.org/wiki/Hash_function#Multiplicative_hashing)

Reply:
@MartinNowak I share your hope that this will be subject to further examination. I changed to this in April (https://github.com/dlang/dmd/pull/8175) because it was a small improvement that could pass quickly through the approval pipeline, not because it was the final word. In particular I'd like if we could use compile-time information about the types and order of the fields to make better choices.